src/main/java/net/minecraftforge/resource/ResourceCacheManager.java

```diff
@@ -0,0 +1,375 @@
+import org.slf4j.Logger;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.NoSuchFileException;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.BiFunction;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Cache manager for resources.
+ * <p>
+ * This class handles caching the resource listing calls on a pack, pack type and namespace level.
+ */
+public class ResourceCacheManager {
+
+    /**
+     * Indicates if the underlying namespaced managers need to support reloading.
+     * Disabled for the (in-jar / downloaded) default vanilla pack.
+     */
+    private final boolean supportsReloading;
+    /**
+     * Indicates if the indexing of the file-tree should happen off-thread or on-thread and block the process accordingly.
+     */
+    private final ForgeConfigSpec.BooleanValue indexOffThreadConfigOption;
+
+    /**
+     * The path builder (different users have different requirements for how they want to handle this)
+     */
+    private final BiFunction<PackType, String, Path> pathBuilder;
+    /**
+     * The individual sub managers by pack type and namespace
+     */
+    private final Map<PackTypeAndNamespace, NamespacedResourceCacheManager> managersByNamespace = new ConcurrentHashMap<>();
+
+    /**
+     * Creates a new instance of a resource cache manager.
+     *
+     * @param supportsReloading True to make the namespace specific managers support reloading the cache.
+     * @param indexOffThreadConfig    True to index the file-tree off-thread.
+     * @param pathBuilder       The path builder to use.
+     */
+    public ResourceCacheManager(final boolean supportsReloading, final ForgeConfigSpec.BooleanValue indexOffThreadConfig, final BiFunction<PackType, String, Path> pathBuilder) {
+        this.supportsReloading = supportsReloading;
+        this.indexOffThreadConfigOption = indexOffThreadConfig;
+        this.pathBuilder = pathBuilder;
+    }
+
+    public boolean shouldIndexOffThread() {
+        return !getConfigValue(this.indexOffThreadConfigOption);
+    }
+
+    public static boolean shouldUseCache() {
+        return getConfigValue(ForgeConfig.COMMON.cachePackAccess);
+    }
+
+    private static boolean getConfigValue(Supplier<Boolean> configValue) {
+        //Yes we catch the early loading error on purpose. This feature needs to be configurable but Mojang loads resources really early.
+        //So when they are loaded early we preserve the default behaviour which is to process it on the main thread.
```

Comment:
Because this method is also used to determine whether to cache pack resource accesses, this comment should be updated to reflect that.